internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -44,75 +47,52 @@ func (k *Kubeadm) ID(data *userdata.UserData) string {
 }
 
 // PreFunc implements the Service interface.
-// nolint: gocyclo
 func (k *Kubeadm) PreFunc(data *userdata.UserData) (err error) {
 	if err = writeKubeadmConfig(data); err != nil {
 		return err
 	}
 
+	// Write out all certs we've been provided
+	if err = writeKubeadmPKIFiles(data); err != nil {
+		return err
+	}
+
+	// Run kubeadm alpha phase certs
+	// this should fill in whatever gaps we have in the provided
+	// certs
 	if data.IsBootstrap() {
-		// Write out all certs we've been provided
-		certs := []struct {
-			Cert     *x509.PEMEncodedCertificateAndKey
-			CertPath string
-			KeyPath  string
-		}{
-			{
-				Cert:     data.Security.Kubernetes.CA,
-				CertPath: constants.KubeadmCACert,
-				KeyPath:  constants.KubeadmCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.SA,
-				CertPath: constants.KubeadmSACert,
-				KeyPath:  constants.KubeadmSAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.FrontProxy,
-				CertPath: constants.KubeadmFrontProxyCACert,
-				KeyPath:  constants.KubeadmFrontProxyCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.Etcd,
-				CertPath: constants.KubeadmEtcdCACert,
-				KeyPath:  constants.KubeadmEtcdCAKey,
-			},
-		}
 
-		for _, cert := range certs {
-			if cert.Cert == nil {
-				continue
-			}
-			if err = writeKubeadmPKIFiles(cert.Cert, cert.CertPath, cert.KeyPath); err != nil {
-				return err
-			}
-		}
-	} else if data.IsControlPlane() {
+		log.Println("placeholder for kubeadm alpha phase cert")
+	}
+
+	// Handle cert propagation via Trustd
+	if data.IsControlPlane() && !data.IsBootstrap() {
 		if data.Services.Trustd == nil || data.Services.Trustd.BootstrapNode == "" {
 			return nil
 		}
 
-		creds, err := basic.NewCredentials(data.Services.Trustd)
-		if err != nil {
-			return err
-		}
-
-		files := []string{
-			constants.AuditPolicyPathInitramfs,
-			constants.EncryptionConfigInitramfsPath,
-		}
+		// Get a list of files we need to request
+		// ( filtered by ones we already have )
+		filelist := files()
 
-		conn, err := basic.NewConnection(data.Services.Trustd.BootstrapNode, constants.TrustdPort, creds)
-		if err != nil {
+		// Initialize trustd peer client connection
+		var trustds []proto.TrustdClient
+		var err error
+		if trustds, err = createTrustdClients(data); err != nil {
 			return err
 		}
 
-		client := proto.NewTrustdClient(conn)
+		var wg sync.WaitGroup
+		for _, fileRequest := range filelist {
+			// Have a single chan shared across all clients
+			// for a given file
+			content := make(chan []byte)
 
-		if err := writeFiles(client, files); err != nil {
-			return err
+			// TODO add ability to cancel other requests
+			go fetchTrustdFiles(fileRequest, &wg, trustds, content)
```

Comment:
I think this can be simplified and easier to reason about with an inline go routine:

go func(r *proto.ReadFileRequest{}, clients []proto.TrustdClient){
    content := fetchTrustdFiles(r, clients)
    writeTrustdFiles(r.Path, content) 
}(fileRequest, trustds)